torchx/lib/torchx/backend.ex

```diff
@@ -0,0 +1,218 @@
+  defp to_blob(%T{data: %TB{ref: ref}} = tensor, nil), do: NIF.to_blob(ref)
+  defp to_blob(%T{data: %TB{ref: ref}} = tensor, limit), do: NIF.to_blob(ref, limit)
+
+  @impl true
+  def backend_deallocate(%Nx.Tensor{data: %TB{ref: ref}}), do: NIF.delete_tensor(ref)
+
+  @impl true
+  def backend_transfer(tensor, Nx.Tensor, opts) do
+    backend_transfer(tensor, Nx.BinaryBackend, opts)
+  end
+
+  def backend_transfer(tensor, Torchx.Backend, _opts) do
+    tensor
+  end
+
+  def backend_transfer(tensor, backend, opts) do
+    backend.from_binary(tensor, to_blob(tensor), opts)
+  end
+
+  @impl true
+  def from_binary(%T{type: type, shape: shape} = out, binary, _opts) do
+    NIF.from_blob(binary, shape, torch_type(type)) |> from_ref(out)
+  end
+
+  ## Shape
+
+  @impl true
+  def reshape(out, %T{data: %TB{ref: ref}} = tensor, shape),
+    do: NIF.reshape(ref, shape) |> from_ref(tensor)
+
+  @impl true
+  def as_type(%T{type: type} = out, %T{data: %TB{ref: ref}} = tensor),
+    do: NIF.to_type(ref, torch_type(type)) |> from_ref(out)
+
+  @impl true
+  def squeeze(out, tensor, _axes) do
+    NIF.squeeze(tensor.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def transpose(out, tensor, axes), do: IO.inspect(axes)
+  # do: NIF.transpose(tensor, dim0, dim1) |> from_ref(out)
+
+  ## Ops
+
+  @impl true
+  def add(out, left, right) do
+    NIF.add(left.data.ref, right.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def dot(
+        out,
+        %T{type: t1, data: %TB{ref: left_ref}} = left,
+        _axes1,
+        %T{type: t2, data: %TB{ref: right_ref}} = right,
+        _axes2
+      ) do
+    NIF.dot(left_ref, right_ref) |> from_ref(out)
+  end
+
+  @impl true
+  def cholesky(%T{type: output_type, shape: {rows, cols}} = out, tensor) do
+    NIF.cholesky(tensor.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def qr(
+        {q_holder, r_holder},
+        tensor,
+        opts
+      ),
+      do: NIF.qr(tensor.data.ref, opts[:mode] == :reduced) |> from_pair_ref({q_holder, r_holder})
+
+  @big_tensor_threshold 10_000_000
+
+  @impl true
+  def inspect(%T{type: {_, elem_size}} = tensor, inspect_opts) do
+    limit = if(inspect_opts.limit == :infinity, do: nil, else: inspect_opts.limit + 1)
+
+    if on_cpu?(tensor) do
+      byte_size = nbytes(tensor)
+      byte_limit = limit && limit * div(elem_size, 8)
+
+      if min(byte_limit, byte_size) > @big_tensor_threshold do
+        "Torchx tensor is too large to inspect. Explicitly transfer the tensor by calling Nx.backend_transfer/1"
+      else
+        alias Inspect.Algebra, as: IA
+
+        binary = to_blob(tensor, limit)
+        result = Nx.Backend.inspect(tensor, binary, inspect_opts)
+        IA.concat(["Torchx.Backend", IA.line(), result])
+      end
+    else
+      raise "Tensor is located on #{device(tensor)} device, so direct to_binary/1 operation is not supported. " <>
+              "You must first transfer the tensor to Elixir by calling Nx.backend_transfer/1"
```

Comment:
Same as above, make sure this returns a string instead of raising:

```suggestion
              "Tensors on the GPU cannot be inspected. Explicitly transfer the tensor by calling Nx.backend_transfer/1"
```